design/ceph/node-loss-rbd-cephfs.md

```diff
@@ -0,0 +1,150 @@
+
+- Remove the volume attachment from the lost node
+- Delete the old pod on the lost node
+- Create a new pod on the new node
+- Allow the volume to be attached to the new node
+
+Once this taint is applied manually, Rook will create a [NetworkFence CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md). The [csi-addons operator](https://github.com/csi-addons/kubernetes-csi-addons) will then blocklist the node to prevent any ceph rbd/CephFS client on the lost node from writing any more data.
+
+After the new pod is running on the new node and the old node which was lost comes back, Rook will delete the [NetworkFencr CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md).
+
+example of taint to be applied to lost node:
+
+```console
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoExecute
+# or
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule
+```
+
+> Note: This will be enabled by default in Rook if the NetworkFence CR is found, in the case for some reason user wants to disable this feature in Rook can edit `Rook-ceph-operator-config` configmap and update the `Rook_WATCH_FOR_NODE_FAILURE: "false"`.
+
+## How to get which IP to blocklist
+
+There are multiple networking options available for example, Host Networking, Pod networking, Multus etc. This make it difficult to know which NodeIP address to blocklist.
+For this we'll follow the following approach which will work for all networking options(but still will not work for external cluster which NodeIPs):
+
+1. Get the `volumesInUse` from the node which has the taint `out-of-service`.
+
+2. List all the pv and compare the pv `spec.volumeHandle` with the node `volumesInUse` field `volumeHandle`
+   1. Example:
+
+      Below is sample Node volumeInUse field
+
+    ```
+      volumesInUse:
+        - kubernetes.io/csi/rook-ceph.rbd.csi.ceph.com^0001-0009-Rook-ceph-0000000000000002-24862838-240d-4215-9183-abfc0e9e4002
+
+      # Note: The volumeInUse naming convention are `kubernetes.io/csi/ + CSI driver name + ^ + volumeHandle`
+    ```
+
+     and the following is pv `volumeInHandle`
+
+     ```
+      volumeHandle: 0001-0009-rook-ceph-0000000000000002-24862838-240d-4215-9183-abfc0e9e4002
+    ```
+
+3. For Ceph volumes on that node:
+
+   1. If RBD PVC makes use of the rbd status API
+      1. example:
+
+    ```console
+    $ rbd status <poolname>/<image_name>
+    Watchers:
+    watcher=172.21.12.201:0/4225036114 client.17881 cookie=18446462598732840961
+    ```
+
+   2. If CephFS PVC uses below CLI to clients connect to subvolume
+      1. example:
+
+      ```console
+      $ ceph tell mds.0 client ls
```

Comment:
This needs to be called for each mds daemon, right? And it could be a long list of clients to compare if a large cluster? This seems fine since it's a rare operation, just wanted to confirm.

Reply:
i think we need to call this with one of the active mds.

Reply:
we can use `ceph tell mds.* client ls` to run for all the daemon.

Yes, it is a large set of outputs. I will like to do something in go-ceph so that we can get these details. But it will time for me to understand and all, meantime I'll start with running the command only .

Reply:
>Yes, it is a large set of outputs. I will like to do something in go-ceph so that we can get these details. But it will time for me to understand and all, meantime I'll start with running the command only .

I would suggest using CLI here (if we don't have an API in go-ceph yet), You will not get much benefit using go-ceph, this is a rare one-time operation when the node goes down and is required only if no rbd PVC but cephfs PVC is mounted on the node.